### ISSUE TITLE
CS8509 Doesn't flag nullable enums correctly

### ISSUE THREAD

**Issue by Euan-McVie** [dotnet/roslyn#71508]
**Version Used**: 8.0.100

**Steps to Reproduce**:

See [sharplab Repro](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKAPQGYACAUxgFcxSAVcgZwBcBRa2gbwNN9IEE4PPgCECAXwIEAxAAcAThADmkUgHcI8mLCWkEURhGAAbcqQDCAZQAcAVkwAWImSwXS3fH1Ilv2AAykrAAesuQAxswAcgD2AOqa2jBKABQMLOw0APykAG4QxgCU7sJevOgA7Ln5pIxqUMxhABbFnqVt1MbGpAC8AHykAER0jQakox3GA0KtbV5pbBwAdPy8fYPDo6P8UyWzvPMZYIsiPf1DI4xjlyIDu17iANy7kjOkuz7o/oEh4czxWjpUkwFlkqoUWnsKmCanUGs0PHs5sDDstVmcNpctjtXnsDksTmtzptrrccXxHs8JDIFMpVBoAUlSPJgdFmRYbPYHEA==)
```cs
public enum TestEnum {
    A,
    B
}

#pragma warning disable CS8524
public class C {
    public string ExpectNoWarning(TestEnum? val) {
        return val switch {
            null => "This is null",
            TestEnum.A  => "This is A",
            TestEnum.B => "This is B"
        };
    }
    
    public string ExpectWarning(TestEnum? val) {
        return val switch {
            TestEnum.A  => "This is A",
            TestEnum.B => "This is B"
        };
    }
}
#pragma warning restore CS8524
```

Note that I have disabled CS8524 as my understanding is that it is primarily to guard against unexpected values and for this case `null` is an expected value due to the parameter being defined as nullable.

**Diagnostic Id**: CS8509

If this is a report about a bug in an analyzer, please include the diagnostic if possible (e.g. `"IDE0030"`).

**Expected Behavior**:

For nullable enums the behaviour of CS8509 should flag if either a declared enum value is missing **or** if the `null` pattern is missing. If all defined enum values **and** `null` are present then it should be treated as a complete set.

**Actual Behavior**:

1. When null is defined CS8509 states that the catch all `_` is not covered.
2. When null is not defined CS8509 is not raised at all.

**Comment by jcouv** [comment]
It is correct that both `switch` expressions report a non-exhaustiveness warning, as neither handles unnamed enum values such as `(TestEnum)2`.

However, the warning given in the first switch (`warning CS8509: The switch expression does not handle all possible values of its input type (it is not exhaustive). For example, the pattern '_' is not covered.`) is less informative than that given in the second switch (`warning CS8524: The switch expression does not handle some values of its input type (it is not exhaustive) involving an unnamed enum value. For example, the pattern '(TestEnum)2' is not covered.`).

It would be nice to improve the diagnostic reported in the first case.

**Comment by mb-ps** [comment]
The ActualBehaviour no 1 is mitigated if null is not given as the first alternative, e.g.
```
    public string ExpectNoWarning(TestEnum? val) {
        return val switch {     
            TestEnum.A  => "This is A",
             null => "This is null",  //Moved this line
            TestEnum.B => "This is B"
        };
    }
```

**Comment by mb-ps** [comment]
I agree with the authors view and I feel that this should be corrected.
